README.pod
=pod

=encoding UTF-8

=for :stopwords Ferenc Erki backend CPAN sed

=head1 NAME

Rex::Hook::File::Diff - show diff of changes for files managed by Rex

=head1 VERSION

version v0.4.0

=head1 SYNOPSIS

    use Rex::Hook::File::Diff;

=head1 DESCRIPTION

This module allows L<Rex> to show a diff of changes for the files managed via its built-in L<file manipulation commands|https://metacpan.org/pod/Rex::Commands::File> which rely on the L<file|https://metacpan.org/pod/Rex::Commands::File#file> command as a backend:

=over 4

=item L<file|https://metacpan.org/pod/Rex::Commands::File#file>

=item L<delete_lines_matching|https://metacpan.org/pod/Rex::Commands::File#delete_lines_matching>

=item L<delete_lines_according_to|https://metacpan.org/pod/Rex::Commands::File#delete_lines_according_to>

=item L<append_if_no_such_line|https://metacpan.org/pod/Rex::Commands::File#append_if_no_such_line>

=item L<append_or_amend_line|https://metacpan.org/pod/Rex::Commands::File#append_or_amend_line>

=item L<sed|https://metacpan.org/pod/Rex::Commands::File#sed>

=back

It prefers to use the C<diff> utility on non-Windows managed endpoints, if available.

=head1 DIAGNOSTICS

This module does not do any error checking (yet).

=head1 CONFIGURATION AND ENVIRONMENT

This module does not require any configuration, nor does it use any environment variables.

=head1 DEPENDENCIES

See the included C<cpanfile>.

Requires the C<diff> utility to show the diff for remote file operations.

=head1 INCOMPATIBILITIES

There are no known incompatibilities with other modules.

=head1 BUGS AND LIMITATIONS

There are no known bugs. Make sure they are reported.

Upload hook support is not implemented (yet), so diff is not shown upon file uploads when using the C<source> option with the L<file|https://metacpan.org/pod/Rex::Commands::File#file> command (or the L<upload|https://metacpan.org/pod/Rex::Commands::Upload#upload> command directly).

=head1 AUTHOR

Ferenc Erki <erkiferenc@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2020,2021 by Ferenc Erki.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Changes
Revision history for Rex-Hook-File-Diff

{{$NEXT}}
 [API CHANGES]

 [BUG FIXES]

 [DOCUMENTATION]

 [ENHANCEMENTS]

v0.4.0 2021-08-18T18:16:37Z
 - No code changes since v0.3.0.4 trial release

v0.3.0.4 2021-08-02T17:45:44Z (TRIAL RELEASE)
 [BUG FIXES]
 - Fix leading message in expected output on Solaris

 [DOCUMENTATION]
 - Fix license typos

v0.3.0.3 2021-08-01T10:23:33Z (TRIAL RELEASE)
 [BUG FIXES]
 - Fix new line in expected output on Solaris

v0.3.0.2 2021-07-31T19:28:19Z (TRIAL RELEASE)
 [BUG FIXES]
 - Fix optional timestamps in headers
 - Fix expected hunk headers on Solaris
 - Fix expected output for identical files on Solaris

v0.3.0.1 2021-07-29T16:56:12Z (TRIAL RELEASE)
 [BUG FIXES]
 - Fix parallel tests
 - Detect managed OS instead of running OS

 [ENHANCEMENTS]
 - Test on Mac OS X with Travis CI
 - Test with GitHub Actions instead of Travis CI
 - Support file command with source option

v0.3.0 2020-08-28T19:04:55Z
 - No code changes since v0.2.0.1 trial release

 [DOCUMENTATION]
 - Add contributor

v0.2.0.1 2020-08-19T17:01:12Z (TRIAL RELEASE)
 [BUG FIXES]
 - Fix tests on Windows

 [ENHANCEMENTS]
 - Use built-in Rex logic to get temp file names
 - Test on Windows with Travis CI

v0.2.0 2020-06-03T09:10:16Z
 - No code changes since v0.1.0.2 trial release

v0.1.0.2 2020-06-02T13:00:05Z (TRIAL RELEASE)
 [BUG FIXES]
 - Mark header timestamps optional

v0.1.0.1 2020-06-01T07:36:34Z (TRIAL RELEASE)
 [ENHANCEMENTS]
 - Use the diff utility if available
 - Use say instead of print

v0.1.0 2020-05-26T17:42:17Z
 - No code changes since v0.0.0.3 trial release

v0.0.0.3 2020-05-19T20:23:50Z (TRIAL RELEASE)
 [ENHANCEMENTS]
 - Add missing test names
 - Add tests for expected set of testing files

v0.0.0.2 2020-05-17T05:45:18Z (TRIAL RELEASE)
 [BUG FIXES]
 - Fix minimum Perl version required

v0.0.0.1 2020-05-16T16:26:26Z (TRIAL RELEASE)
 - Initial release


lib/Rex/Hook/File/Diff.pm
package Rex::Hook::File::Diff;

# ABSTRACT: show diff of changes for files managed by Rex

use 5.012;
use warnings;

use File::Basename;
use Rex 1.013004 -base;
use Rex::Helper::Run;
use Rex::Hook;
use Text::Diff 1.44;

our $VERSION = '9999';

register_function_hooks { before_change => { file => \&show_diff, }, };

sub show_diff {
    my ( $original_file, @opts ) = @_;

    my $diff;
    my $diff_command = can_run('diff');
    state $managing_windows = is_windows();

    if ( !$managing_windows && $diff_command ) {
        my $command = join q( ), $diff_command, '-u', involved_files($original_file);

        if ( $diff = i_run $command, fail_ok => TRUE ) {
            $diff .= "\n";
        }
    }
    elsif ( Rex::is_local() ) {
        $diff = diff( involved_files($original_file) );
    }
    else {
        Rex::Logger::debug(
            'Skipping File::Diff hook due to remote operation without the diff utility');
        return;
    }

    if ( length $diff > 0 ) {
        Rex::Commands::say("Diff for: $original_file\n$diff");
    }

    return;
}

sub involved_files {
    my $file = shift;

    my $temp_file = Rex::Commands::File::get_tmp_file_name($file);
    my $null      = File::Spec->devnull();

    if ( !is_file($file) )      { $file      = $null } # creating file
    if ( !is_file($temp_file) ) { $temp_file = $null } # removing file

    return ( $file, $temp_file );
}

1;

__END__

=for :stopwords backend CPAN sed

=head1 SYNOPSIS

    use Rex::Hook::File::Diff;

=head1 DESCRIPTION

This module allows L<Rex> to show a diff of changes for the files managed via its built-in L<file manipulation commands|https://metacpan.org/pod/Rex::Commands::File> which rely on the L<file|https://metacpan.org/pod/Rex::Commands::File#file> command as a backend:

=over 4

=item L<file|https://metacpan.org/pod/Rex::Commands::File#file>

=item L<delete_lines_matching|https://metacpan.org/pod/Rex::Commands::File#delete_lines_matching>

=item L<delete_lines_according_to|https://metacpan.org/pod/Rex::Commands::File#delete_lines_according_to>

=item L<append_if_no_such_line|https://metacpan.org/pod/Rex::Commands::File#append_if_no_such_line>

=item L<append_or_amend_line|https://metacpan.org/pod/Rex::Commands::File#append_or_amend_line>

=item L<sed|https://metacpan.org/pod/Rex::Commands::File#sed>

=back

It prefers to use the C<diff> utility on non-Windows managed endpoints, if available.

=head1 DIAGNOSTICS

This module does not do any error checking (yet).

=head1 CONFIGURATION AND ENVIRONMENT

This module does not require any configuration, nor does it use any environment variables.

=head1 DEPENDENCIES

See the included C<cpanfile>.

Requires the C<diff> utility to show the diff for remote file operations.

=head1 INCOMPATIBILITIES

There are no known incompatibilities with other modules.

=head1 BUGS AND LIMITATIONS

There are no known bugs. Make sure they are reported.

Upload hook support is not implemented (yet), so diff is not shown upon file uploads when using the C<source> option with the L<file|https://metacpan.org/pod/Rex::Commands::File#file> command (or the L<upload|https://metacpan.org/pod/Rex::Commands::Upload#upload> command directly).

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.024.
use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "show diff of changes for files managed by Rex",
  "AUTHOR" => "Ferenc Erki <erkiferenc\@gmail.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Rex-Hook-File-Diff",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.014",
  "NAME" => "Rex::Hook::File::Diff",
  "PREREQ_PM" => {
    "File::Basename" => 0,
    "Rex" => "1.013004",
    "Rex::Helper::Run" => 0,
    "Rex::Hook" => 0,
    "Text::Diff" => "1.44",
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "English" => 0,
    "File::Temp" => 0,
    "File::Touch" => "0.08",
    "Rex::Commands::File" => "1.012",
    "Test2::V0" => "0.000071",
    "Test::File" => "1.443",
    "Test::Output" => "0.03",
    "re" => 0
  },
  "VERSION" => "v0.3.0.4",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "English" => 0,
  "File::Basename" => 0,
  "File::Temp" => 0,
  "File::Touch" => "0.08",
  "Rex" => "1.013004",
  "Rex::Commands::File" => "1.012",
  "Rex::Helper::Run" => 0,
  "Rex::Hook" => 0,
  "Test2::V0" => "0.000071",
  "Test::File" => "1.443",
  "Test::Output" => "0.03",
  "Text::Diff" => "1.44",
  "re" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
